### PR TITLE
CRM-20392: remove cruft code and added unit test

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20392: Credit card payment not recorded when contribution is partially paid](https://issues.civicrm.org/jira/browse/CRM-20392)

### CODE REVIEW

**Comment by monishdeb** [comment]
@eileenmcnaughton I have pulled out remaining changes from #10126 (closed it) here

File: tests/phpunit/CRM/Contribute/Form/ContributionTest.php [hunk 1]
```diff
@@ -1106,4 +1105,55 @@ public function testCardTypeAndPanTruncationLiveMode() {
...
+      'credit_card_type' => 'Visa',
+      'billing_city-5' => 'Vancouver',
+      'billing_state_province_id-5' => 1059,
+      'billing_postal_code-5' => 1321312,
+      'billing_country_id-5' => 1228,
+      'trxn_date' => '2017-04-11 13:05:11',
+    ), 'live');
+    $contribution = $this->callAPISuccessGetSingle('Contribution', array());
+    $this->assertNotEmpty($contribution);
+    $this->assertEquals('Completed', $contribution['contribution_status']);
```
Comment (monishdeb) [inline tests/phpunit/CRM/Contribute/Form/ContributionTest.php]: @eileenmcnaughton I think we need to comment this line for the moment, as its strange that this scenario works well via UI

Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/Form/ContributionTest.php]: It just passed for me locally - does it pass in isolation for you

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by seamuslee001** [comment]
Needs rebasing

**Comment by eileenmcnaughton** [comment]
I think this will fix that test #10286

I did discover another possible issue in the process caused by a JMA fix - I have just added a PR to signpost it #10287

File: tests/phpunit/CRM/Contribute/Form/ContributionTest.php (hunk 1 above)
Comment (monishdeb) [inline tests/phpunit/CRM/Contribute/Form/ContributionTest.php]: yup it passed earlier in my local 
```
Monishs-MacBook-Pro:civicrm monish$ phpunit4 --filter=testPartialPaymentWithCreditCard tests/phpunit/CRM/Contribute/Form/ContributionTest.php
PHPUnit 4.8.21 by Sebastian Bergmann and contributors.

.
Installing civicrm47_drupal_test database


Time: 6.17 seconds, Memory: 67.00Mb

OK (1 test, 10 assertions)
```
but unfortunately, it doesn't get passed in jenkin... don't knw why

**Comment by monishdeb** [comment]
@seamuslee001 @eileenmcnaughton I have rebased the PR as #10286 is merged, but still the test build fails on status match. Should we comment out https://github.com/civicrm/civicrm-core/pull/10285/files#diff-f5dec9266a02875f07e5948eed1ca68bR1156 with proper ```@todo``` message? (as it pass in local)

**Comment by seamuslee001** [comment]
@monishdeb I just cherry-picked this onto a local install and ran 2 separate tests 1 the specific test in isolation and one of the whole of that file and these are the results
```
seamus@civicrmdevelopment7:~/buildkit/build/47.demo/sites/all/modules/civicrm$ ./tools/scripts/phpunit 'CRM_Contribute_Form_ContributionTest'
Installing 47demotest_c6lz0 schema
PHPUnit 4.8.21 by Sebastian Bergmann and contributors.

.
Installing 47demotest_c6lz0 database
......................F

Time: 2.48 minutes, Memory: 54.00Mb

There was 1 failure:

1) CRM_Contribute_Form_ContributionTest::testPartialPaymentWithCreditCard
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'Completed'
+'Partially paid'

/home/seamus/buildkit/build/47.demo/sites/all/modules/civicrm/tests/phpunit/CRM/Contribute/Form/ContributionTest.php:1156
/home/seamus/buildkit/build/47.demo/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:184
/home/seamus/buildkit/bin/phpunit4:545

FAILURES!
Tests: 24, Assertions: 259, Failures: 1.
seamus@civicrmdevelopment7:~/buildkit/build/47.demo/sites/all/modules/civicrm$ ./tools/scripts/phpunit 'CRM_Contribute_Form_ContributionTest' --filter 'testPartialPaymentWithCreditCard'
PHPUnit 4.8.21 by Sebastian Bergmann and contributors.

.
Installing 47demotest_c6lz0 database


Time: 11.05 seconds, Memory: 16.57Mb

OK (1 test, 10 assertions)
```

File: tests/phpunit/CRM/Contribute/Form/ContributionTest.php (hunk 1 above)
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/Form/ContributionTest.php]: It passes if you comment out this line

$this->relationForFinancialTypeWithFinancialAccount($this->_financialTypeId);

in an earlier function - I think there is caching somewhere

THIS

  public static function getTaxRates() {
    if (!isset(Civi::$statics[__CLASS__]['taxRates'])) {
      Civi::$statics[__CLASS__]['taxRates'] = array();

should be OK as Civi::$statics is cleared between test runs - so although that seemed like the obvious place I don't think it is there.

**Comment by eileenmcnaughton** [comment]
See comment at https://github.com/civicrm/civicrm-core/pull/10285#discussion_r114084829 for unit test

**Comment by monishdeb** [comment]
@eileenmcnaughton you are right, ```Civi::$statics[__CLASS__]['taxRates']``` doesn't get flushed out after each unit test which involves 'Sale Tax' evaluation, get executed as a result of which, there always a tax amount set in that ```static``` variable. So I have tried this function https://gist.github.com/monishdeb/59bdf0f01112d53a52666233d838dd9e to do the job at ```tearDown()``` but without any success. So I shifted the new UT definition above (just before the UT which ```enableTaxAndInvoicing()``` and it got passed for obvious reason :D (I know that's not a good practice to squash the bug which is still out there affecting the test env.)

**Comment by monishdeb** [comment]
@seamuslee001 thanks for your feedback and I found the reason and fixed it. Hopefully, the test build will pass this time :)

**Comment by seamuslee001** [comment]
Happy that it passes now @monishdeb are you able to log in JIRA that bug as i think we should really fix that

**Comment by monishdeb** [comment]
@seamuslee001 yup created a backlog issue in JIRA - https://issues.civicrm.org/jira/browse/CRM-20505

**Comment by totten** [comment]
Expectations:
 * `Civi::$statics` should be cleared automatically between tests (by way of `CiviUnitTestCase::setUp()`).
 * Each typical subclass of `CiviUnitTestCase` should do an expensive DB reset *once* (dropping/recreating all sample data when it starts executing).
 * The cleanup logic between test functions varies by class. Some things (like calling `Civi::reset()`) should be universal. But tests can be very different about cleaning up the DB. (Personally, I prefer `$this->useTransaction()` to ensure that DB is cleaned up... but there are trade-offs in every approach.) Guess: `Civi::$statics` is a red-herring. If the underlying *DB content* is not properly reset, then the `Civi::$statics` cache would get rehydrated from bad info.

**Comment by eileenmcnaughton** [comment]
Ok let's merge & punt the other issue out.

**Pull request merged**